Add controllers leader election documentation (#1890)
aws-controllers-k8s/community: docs/content/docs/user-docs/leader-election.md

---
title: "Managing multiple instances of ACK with leader election"
description: "Configure leader election for ACK controllers"
draft: false
menu:
  docs:
    parent: "getting-started"
weight: 60
toc: true
---

In a Kubernetes cluster, you may want to run multiple instances of any ACK
controller - configured for different accounts or regions, or for fail state 
rollover. 
However, to avoid conflicts and ensure proper resource management, it's necessary

> to designate one instanceas the leader, which takes responsibility for executing

Reviewer: typo: `instanceas` -> `instance as`

certain operations while the other instances remain passive. In the event that
the leader instance fails, [leader election][leader-election] ensures the seamless
transition of leadership to another healthy instance.

By default, leader election is disabled in the ACK Helm charts. However, once
enabled, you gain the flexibility to scale the default deployment of ACK controllers
from a single replica (1) to a higher number.

## Enabling Leader Election for ACK Controllers

To enable leader election when installing an ACK controller, set the
`leaderElection.enabled` to `true` in the helm chart `values.yaml` like:

```yaml
leaderElection:
    enabled: true
```

You also have the flexibility to scale the number of controller replicas. Edit
the `deployment.replicas` configuration in the same `values.yaml` file and
adjust it to your desired count, such as:

```yaml
deployment:
    replicas: 3
```

> ## Configuring Leader Election namespace

Reviewer: capitalize `Namespace` in this heading

---

> title: "Leader election"

Author: Hesitant to name this `Managing Controller Leadership`

Reviewer 2: All of our other docs use a verb, so I would prefer it to match that format. My suggestion is: `Managing multiple instances of ACK with leader election`?

description: "Configure leader election for ACK controllers"

> lead: "How to enable leader election for ACK controllers"

Reviewer 2: I don't think this lead adds any additional value, can probably get rid of it.

draft: false
menu:
  docs:
    parent: "getting-started"
weight: 60
toc: true
---

> In a Kubernetes cluster, multiple instances of a controller maybe running
> simultaneously to manage resources and perform tasks. However, to avoid conflicts

Reviewer 2: ```suggestion
In a Kubernetes cluster, you may want to run multiple instances of any ACK controller - configured for different accounts or regions, or for fail state rollover. However, to avoid conflicts
```

and ensure proper resource management, it's necessary to designate one instance
as the leader, which takes responsibility for executing certain operations while
the other instances remain passive. In the event that the leader instance fails,
[leader election][leader-election] ensures the seamless transition of leadership
to another healthy instance.

> In ACK land, by default, the feature of leader election is disabled. To utilize

Reviewer 2: I'm not a fan of `In ACK land` lol. I'd just say `By default, the ACK Helm charts disable leader election`

it, users can configure leader election using either Helm chart values or
controller flags. Once enabled, you have the flexibility to scale the default

> deployment of ACK controllers from a single replica (1) to a higher number.

Reviewer 2: `or controller flags` is implied when using the Helm chart. We don't advise installing directly from manifests, so our docs don't reference those.

## Enabling Leader Election for ACK Controllers

> If you would like to enable leader election for your ACK controllers, set the

Reviewer 2: ```suggestion
To enable leader election when installing an ACK controller, set the
```

> `leaderElection.enabled` to true in the helm chart `values.yaml` like:

Reviewer 2: true -> `true`

```yaml
leaderElection:
    enabled: true
```

You also have the flexibility to scale the number of controller replicas. Edit
the `deployment.replicas` configuration in the same `values.yaml` file and
adjust it to your desired count, such as:

```yaml
deployment:
    replicas: 3

> ```

Author: This doesn't exist. Thinking about adding it. Thoughts @RedbackThomson ?

Reviewer 2: Yes if we support leader election, this should be included

Author: Opened https://github.com/aws-controllers-k8s/code-generator/pull/465/files

## Configuring Leader Election namespace

> The Leader Election Namespace is a controller configuration setting that

Reviewer 2: `Leader Election Namespace` -> `leader election namespace`
